presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ConvertJoinTreeToJoinGraph.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.tree.Expression;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.sql.ExpressionUtils.extractConjuncts;
+
+public class ConvertJoinTreeToJoinGraph
+        implements Rule
+{
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        // We check that join distribution type is absent because we only want to do this transformation once.
+        if (node instanceof JoinNode && ((JoinNode) node).getType().equals(JoinNode.Type.INNER) && !((JoinNode) node).getDistributionType().isPresent()) {
+            return Optional.of(new FlattenedJoinTree((JoinNode) node).toJoinGraphNode(idAllocator));
```

Comment:
For instance (and that is real case that happens):
```
    DistinctLimit
          |
     Join(output: a)
     /                 \
Node              Node
(output: a)      (output: b)
```
after your rewrite:
```
    DistinctLimit
          |
     Join(output: a, b)
     /                 \
Node              Node
(output: a)      (output: b)
```
plan is semantically different. Generally, to be on safe side (since you cannot assume semantics of all nodes), you should preserve output of rewritten node. Presto and Calcite work like that.